### PR TITLE
More pAI Stuff

### PR DESCRIPTION
**Pull request by Geevies**
* pAI can now click on deployed modular computers to turn them on or look a them, they can't interact with them without being in them, though.
* You can now repair pAI with nanopaste.
* pAI can now examine while folded up on the ground, or directly on a person, or inside a computer on the ground or on someone's person.
* pAI now start with the Access Local Computer verb instead of having to relog to gain it when they get put into a computer.
* Airlocks no longer crush mobs that can pass through hatches, instead, the hatch will open to accomodate them. Access does not get taken into account.
* Attacking a pAI card will now attack the pAI inside.

### CODE REVIEW

**Comment by Geevies** [comment]
!review

**Review by Chada-1 — APPROVED** [review]

File: code/modules/examine/examine.dm [hunk 1]
```diff
@@ -82,3 +82,19 @@
...
+	if(client?.eye == src)
+		return TRUE
+	return FALSE
+
+/mob/living/silicon/pai/can_examine()
+	. = ..()
+	if(!.)
+		if(isturf(card.loc)) // are we folded in on the ground
+			return TRUE
+		if(ishuman(loc.loc)) // are we on a human, ie carded in pocket, or in a holder on their head
```
Comment (NonQueueingMatt) [inline code/modules/examine/examine.dm:96]: This check is prone to fucking up if a pAI card is inside a box in a bag or whatever. Theres a recursive helper for this.

File: code/modules/examine/examine.dm [hunk 2]
```diff
@@ -82,3 +82,19 @@
...
+			return TRUE
+		var/obj/item/modular_computer/MC = card.loc
+		if(istype(MC) && (isturf(MC.loc) || ishuman(MC.loc))) // we're inside someone's PDA or laptop, and we're on the ground or on their person
```
Comment (NonQueueingMatt) [inline code/modules/examine/examine.dm:99]: Ditto on the last loc check.

File: code/game/objects/items/devices/paicard.dm [hunk 3]
```diff
@@ -77,6 +77,11 @@
 			EK.forceMove(get_turf(src))
 			installed_encryptionkeys -= EK
 		recalculateChannels()
+	else if(istype(C, /obj/item/stack/nanopaste))
+		if(!pai)
+			to_chat(user, SPAN_WARNING("You cannot repair a pAI device if there's no active pAI personality installed."))
+			return
+		pai.attackby(C, user)
```
Comment (NonQueueingMatt) [inline code/game/objects/items/devices/paicard.dm:84]: Is there a reason to not just reroute all attacks to the pAI mob at the end of the pai card's attackby at all?

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/paicard.dm (hunk 3 above)
Comment (Geevies) [reply code/game/objects/items/devices/paicard.dm:84]: i guess not

**Review by JohnWildkins — APPROVED** [review]

File: code/_helpers/unsorted.dm [hunk 4]
```diff
@@ -548,7 +548,14 @@ Turf and target are seperate in case you want to teleport some distance from a t
...
 	var/datum/progressbar/progbar
 	if (display_progress && user.client && (user.client.prefs.toggles_secondary & PROGRESS_BARS))
-		progbar = new(user, delay, target)
+		var/atom/loc_check = target
+		var/endless_loop_counter = 0
+		while(!isturf(loc_check.loc))
+			loc_check = target.loc
+			endless_loop_counter++
+			if(endless_loop_counter > 5)
+				break
```
Comment (NonQueueingMatt) [inline code/_helpers/unsorted.dm:557]: this is just a for loop:
```
for(var/i = 0; !isturf(loc_check.loc && endless_loop_counter < 5; i++)
	loc_check = target.loc
```

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]
Are all of the manual overrides converted to attack_pai?

File: code/_helpers/unsorted.dm (hunk 4 above)
Comment (Karolis2011) [inline code/_helpers/unsorted.dm:557]:
```suggestion
		for(var/i = 0; !isturf(loc_check.loc) && i < 5; i++)
			loc_check = target.loc
```

**Review by Chada-1 — APPROVED** [review]

**Comment by Geevies** [comment]
> 
> 
> Are all of the manual overrides converted to attack_pai?

yeah, everything not to do with pAI seems to be removed from the card's attackby, otherwise it's passed to it

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**